jest --changedSince=origin/master

alsuren opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. The two fields below are mandatory. -->

<!-- Please remember to update CHANGELOG.md in the root of the project if you have not done so. -->

**Summary**

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

This is a duplicate of https://github.com/facebook/jest/pull/5188 on the off-chance that the CI server's cache per-PR, and is therefore only poisoned for that PR.

I often have jest --watch running in a window, but sometimes I forget to look at it before committing, or I stage a bunch of changes that break tests, and it hides them from the test runner. This means that I often get surprised by CI test failures when I come to open pull requests.

This change allows you to write jest --watch --changedFilesToContributeTo=origin/master to test your entire feature branch, but not the whole project. It also tests only your changes if your branch has diverged from origin/master.

**Test plan**

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI. -->

```
~/src/jest$ yarn jest --listTests | wc -l
     220
~/src/jest$ yarn jest --listTests --changedFilesToContributeTo=origin/master | wc -l
      12
```

SimenB commented:
This needs a rebase.

Also, copying over my comment from the other PR.

>`--changedFilesToContributeTo` is pretty verbose, what about just `--changedSince`?

SimenB commented on CHANGELOG.md:
```diff
@@ -12,6 +12,9 @@
 * `[jest-cli]` Make Jest exit without an error when no tests are found in the
   case of `--lastCommit`, `--findRelatedTests`, or `--onlyChanged` options
   having been passed to the CLI
+* `[jest-cli]` `--changedSince`: allow selectively running tests for code
```
can you move this to just below `master`?

SimenB commented on docs/CLI.md:
```diff
@@ -108,6 +108,12 @@ If you want to inspect the cache, use `--showConfig` and look at the
 Runs tests related to the current changes and the changes made in the last
 commit. Behaves similarly to `--onlyChanged`.
 
+### `--changedSince`
```
Can you add a "Since Jest 22.1.3+" or something?

alsuren replied:
I said 22.2.0 because announcing that a feature is available since a patch release makes me uneasy. If it turns out that we release this feature without bumping the minor version then then `##### available in Jest **22.2.0+**` is not strictly a lie ;-)

SimenB commented:
Question: If I say `--changedSince=HEAD` and then make a commit, would this then know what changed, or do I have to do something like ```--changedSince=`git rev-parse HEAD` ```?

alsuren commented:
Thanks for sticking with me as reviewer while I changed the spec and made a thousand rebases :D 

If you say --changedSince=HEAD then git will re-evaluate what HEAD is every time, so you would indeed need to `git rev-parse HEAD` if you want the base point to stay put.

The flag is designed to so that you can just `jest --watch --changedSince=origin/master` and leave it running in the background forever (even when you switch what you're working on, or other people push things to origin/master) and it will always only test the changes that you're about to contribute.

If you just want to test the changes since the last time you did a git push, I think that `--changedSince=@{upstream}` should do the trick (according to `man git rev-parse`)

jwbay commented:
Just used this for the first time, and it's awesome. Fits my workflow perfectly. Thanks @alsuren!
